File: c/editDistance.c
#include <stdio.h>
#include <string.h>

int min(int x, int y, int z)
{
  return min(min(x, y), z);
}

int editDistance(char *str1, char *str2, int m, int n)
{
  int dp[m+1][n+1];
  for(int i = 0; i <= m; i) {
    for(int j = 0; j <= n; j++) {
      if(i == 0)
        dp[i][j] = j;
      else if(j == 0)
        dp[i][j] = i;
      else if(str1[i-1] == str2[j -1])
        dp[i][j] = dp[i-1][j-1];
      else
        dp[i][j] = 1 + min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]);
    }
  }
  return dp[m][n];
}
int main()
{
  char str1[] = "sunday";
  char str2[] = "saturday";

  int m = strlen(str1);
  int n = strlen(str2);
  printf("edit distance: %d\n", editDistance(str1, str2, m, n));
  return 0;
}

File: ocaml/edit_distance.ml
let rec edit_distance s1 s2 =
  let m = String.length s1 in
  let n = String.length s2 in
  let rec aux i j =
    if i = m || j = n then
      max i j
    else if s1.[i] = s2.[j] then
      aux (i + 1) (j + 1)
    else
      1 + min(aux (i + 1) j) (aux i (j + 1)) (aux(i +1) (j + 1))
  in
  aux 0 0
  ;;
